### ISSUE TITLE
Buttons for CRUD operations from UI

### ISSUE THREAD

**Issue by SwapnilKate** [mrin9/Angular2_SpringBoot#6]
Hi. First of let me appreciate your work as it is truly amazing work which you guys have done. 

This is not an issue but kinda feature I would like to add and know how I can add. Can we have buttons to add, remove , update and delete buttons for each section on UI.

Secondly I tried adding employee from postman but got the error on insert as id cannot be null. Below is the request and response

{ 
   "address1": "string",
   "address2": "string",
   "avatar": "string", 
   "city": "string",  
   "country": "string",  
   "department": "string",  
   "email": "string",
   "firstName": "Swapnil",  
   "id": 300, 
   "jobTitle": "string",
   "lastName": "string",  
   "managerId": 0,
   "phone": "string", 
   "postalCode": "string", 
   "state": "string"
 }

{
  "operationStatus": "ERROR",
  "operationMessage": "NULL not allowed for column \"ID\"; SQL statement:\ninsert into employees (id, address1, address2, avatar, city, country, department, email, first_name, job_title, last_name, manager_id, phone, postal_code, state) values (null, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?) [23502-193]"
}

Can you please help me out for this as I am learning to do SAP with Spring boot and angular 4.

Thanks

**Comment by mrin9** [comment]
Thanks for pointing it out. The issue was that my model class annotated the ID with auto generation therefore it was passing null even when supplied a valid value. I have [commented it now](https://github.com/mrin9/Angular-SpringBoot-REST-JWT/blob/master/src/main/java/com/app/model/employee/Employee.java#L14). It should work fine now.

I am closing this issue. In case you still have trouble, please reopen


To answer can we have buttons to have CRUD operations? yes you can have them, the way I would do it is 

First understand how the project has organized its content. It has two main folder `src` and `webui` src contains the server side code and webui contains the angular code 

under the `webui\src\app` it has set of folders clearly named what it contains. Now that you want to add some buttons for CRUD operation
- First locate the page where you want to add the button
- Most likely you will find that page under  `webui\src\app\pages` folder, identify the page and update its HTML to add a button
- Upon clicking the button it should redirect you to a new URL location
- Create a component (which is basically a page) that accepts some input and upon submit makes an ajax call to the REST server. I normally create a service that makes this REST call and inject it into my component(page).
- Now attach this component to the URL location in `webui\src\app\app-routing.module.ts` so that the component is automatically loaded when you go to the URL. This is what your button is doing
 

I will be happy to add some of your contribution if you decide to go ahead and add a page demonstrating CRUP operation

**Comment by SwapnilKate** [comment]
Hi Sir,

Thank you for your prompt response. Basically I am writing one project in which I need to have Spring Boot with Angular 4 and SPA. I need to understand loads of stuff like how does the SPA happens in spring boot and all although I have knowledge of Spring Boot and Angular.

I am completely new to Webpack and all so was curious how can change javascript and see my changes reflecting. I tried a lot but was not able to proceed.

Basically I wanted to add a new server context and some CRUD operations and see whether everything is getting reflected or not but I couldn't do it.

My background is from ionic and angular1.3 and need to understand this SPring-Boot_Angular SPA in detail. 

It will be beneficial for students like us if you can give us some head starts.

It's really nice to have people like you supporting us on open source. Thanks

**Comment by SwapnilKate** [comment]
Hi,

I would like to know one thing. How do we add to read user info from databse when security is on and after successful authentication. Currently it reads from data.sql but I did not understand the code of reading user details after successful login. Can you please explain about it.

Thank you.

**Comment by mrin9** [comment]
Currently it is reading from Database (H2 Database) not from data.sql. data.sql is only used during the startup to populate the H2 database with some sample data. Once populated it has no use. All the REST services are reading it from H2 database.  If you have setup the project as per instructions you can access this database at [http://localhost:9119/h2-console](http://localhost:9119/h2-console)

the credentials are 
```
Driver    = org.h2.Driver 
JDBC URL  = jdbc:h2:mem:demo 
User Name = sa
```
try changing some data and login with that data, you will notice that it uses the changed data not the data from data.sql

**Comment by SwapnilKate** [comment]
Thank you for your response.

I am currently using your project as a reference and I would like to have on proxy-config to avoid cors issue. I am not able to hit the proxy-config. Can you help us to have the proxy config file to avoid CORS issue and save the development time.

Thank you.

**Comment by mrin9** [comment]
CORS is all-ready taken care of [here](https://github.com/mrin9/Angular-SpringBoot-REST-JWT/blob/master/src/main/java/com/app/config/CorsFilter.java)

So shouldn't be having any CORS issue.  If you still face any problem go ahead and change the above code to suite your needs

**Comment by SwapnilKate** [comment]
Hi Sir,

Thank you for your prompt response. Problem here is we had an older spring security which was form based security and in that we did not implement the CORS part so I have not used your CORS filter.

Secodly I want to run the client code on 4200 port and server on 8090 port so client development can be done very fast. But when I fire any API from 4200 it gives me the CORS error.

I have written the below proxy.config method in your code to bypass the CORS.

{
  "*/":{
    "target": "http://localhost:8090",
    "secure" : false,
    "loglevel": "debug",
    "changeOrigin": true,
    "pathRewrite":{"^/":""}
  }
}
Still when the login request return me the json output I am not able to move forward on home page because of the CORS issue.

I hope I am able to convey you my problem.

Secondly I again would like to thank you for this amazing support from you. 

May I request you for your email id sir because we are looking out for some Full stack consultants for our startups in angular 4 and spring. Looking at your work we all are amazed. You can contact me on my below email id sir.

swapskate@gmail.com

Thank you.

**Comment by SwapnilKate** [comment]
Hi @mrin9 . Thank you for your help. I was finally able to resolve the proxy config error.

Now one more problem I am facing is of context-path:

If I add server.context-path as "/NBU" I get 404 for all js files and css. Where should I add this context variable so that all js and css files should add to their path.

Is there any other method to handle this server context path resolution.

Thank you.
